resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala

```diff
@@ -815,6 +815,7 @@ private[spark] class ApplicationMaster(
       case Shutdown(code) =>
         exitCode = code
         shutdown = true
+        allocator.setShutdown(true)
```

Comment:
so maybe I'm missing a call but it looks like this  Shutdown message is only sent in Client mode.  Is that the only time you are seeing this issue?  ie when you have the driver is local and the applicationmaster and yarn allocator are on the cluster?
But the log message in the description has YarnClusterSchedulerBackend which makes me think this is cluster mode.

Reply:
> it looks like this Shutdown message is only sent in Client mode

Yea, I just noticed that, do you think it's a good idea to send `Shutdown` in cluster mode as well? or any other suggestions? cc @AngersZhuuuu as you are the author of that code.

> the log message in the description has YarnClusterSchedulerBackend which makes me think this is cluster mode.

You are right, my reported job failed in cluster mode, and I think both yarn client/cluster modes have this issue.

Reply:
yeah this won't fix your issue as is so we would need something for cluster mode. I'm fine with sending the shutdown message.  It would be nice if you could test the fix on your use case to make sure it fixes it as well.

Reply:
It's ok to send `Shutdown` in cluster mode too, not only for client mode.